src/rust/engine/fs/store/src/lib.rs

```diff
@@ -267,11 +268,59 @@ impl RemoteStore {
           }
           local_store
             .store_bytes(entry_type, digest.hash, bytes, true)
-            .await?;
-        }
+                .await?;
+            }
+            Ok(())
+        })
+        .await
+    }
+
+    async fn download_digest_to_local_batch(
+        &self,
+        local_store: local::ByteStore,
+        digests: Vec<Digest>,
+    ) -> Result<(), StoreError> {
+        let (large_files, small_files): (Vec<Digest>, Vec<Digest>) = digests
+            .into_iter()
+            .partition(|d| ByteStore::should_use_fsdb(EntryType::File, d.size_bytes));
+
+        let large_file_futures = large_files.into_iter().map(|digest| {
+            let local_store = local_store.clone();
+            let remote_store = self.store.clone();
+
+            async move {
+                local_store
+                    .get_file_fsdb()
+                    .write_using(digest.hash, |file| {
+                        Self::remote_writer(&remote_store, digest, file)
+                    })
+                    .await
+            }
+        });
+
+        let mut local_store = local_store.clone();
+        let small_file_future = async move {
+            let hashmap = self
+                .store
+                .load_bytes_batch(small_files, &mut local_store)
+                .await
+                .map_err(|e| StoreError::Unclassified(e.to_string()))?;
+
+            let missing_digests = hashmap.iter().find(|(_, r)| r.is_err());
+            if let Some((d, r)) = missing_digests {
+                return Err(StoreError::MissingDigest(
```

Comment:
In the event a digest fails to download, this will bubble up the first occurrence from a batch

Ultimately this influences the behavior of `check_cache_content` https://github.com/pantsbuild/pants/blob/22522515efc1753b3d760cbefb80bd65500c182c/src/rust/engine/process_execution/src/lib.rs#L978